### ISSUE TITLE
Please provide two features to realize "copy-on-write"

### ISSUE THREAD

**Issue by ygc369** [dotnet/coreclr#1792]
If some "half-immutable" objects can copy on write, a lot of memory would be saved. 
To realize this, two features are needed.
(1)  I need to get the reference count (rc) of an object to decide whether to copy. If rc is one, there's no need to copy, modifying the object directly is safe. If rc is more than one, we have to copy the object first, and then modify the new object. (all above is what "half-immutable" means.)
for example:
[RC]  //reference count of  objects of class with attribute "RC" would be tracked by CLR.
class test 
{
  int a;
  ~test() //finalizer
  {
    ......
  } 
}
test c = new test();
int rc=RC.get(c); //rc=1
test b=c;
rc = RC.get(c); //rc=2, if b is not used any more in the later, then rc may be 1 here, especially in "release mode".
b=null;
rc= RC.get(c); //rc=1
c=null; //rc=0, the finalizer should be called at once and wouldn't be called when GC happens.

(2)  Updating the reference of an object in the object's non-static member functions should be allowed.
for example:
[RC]class test
{
    int a;
    public void add(int x)
    {
        if(RC.get(this)>1)this=new test(); //should be allowed
        a+=x; //"a" here may be the field of the new object
    }
}

test c =new test();//c.a=0
c.add(1); //c.a=1,no new object allocated
test b=c;//the references of b and c are same, b.a=1
c.add(1);//the reference of c changed, the reference of b not changed, c!=b, c.a=2, b.a=1

**Comment by svick** [comment]
Do you have any ideas about how this could be implemented efficiently? I can't think of anything better than performing a garbage collection every time you call `RC.get()`, which I think would eliminate any performance gains from making your objects "half-immutable" instead of fully immutable.

Have you considered a design similar to immutable collections instead? There, objects are mutable at first, but then they are frozen. When you try to mutate an object that hasn't been frozen yet, it's mutated directly (similar to your suggestion when rc = 1), but when it's frozen, a copy is created (similar to rc > 1).

**Comment by ygc369** [comment]
@svick 
There's no need to perform GC every time "RC.get()" called. A hidden field "reference_count" should be put into this kind of class, and JIT/CLR should maintain it automatically.
There's another usage of reference count. If a class with rc has finalizer, then the finalizer should be called immediately and automatically when rc becomes zero. Thus, GC would be accelerated.

**Comment by OtherCrashOverride** [comment]
This will not work as there is currently not a way to decrement the reference count when a variable simply goes out of scope rather than being assigned null.
```
void sample()
{
  var a = new test();
  {
     var b = a; // refcount++
  }
  // "b" is out of scope so "a" needs to decrement the ref count

  int rc = a.RefCount(); // this incorrectly reports 2 instead of 1
}

**Comment by ygc369** [comment]
@OtherCrashOverride 
JIT/CLR should know when a variable goes out of scope, so the reference count can be updated. This feature need CLR's help.

**Comment by OtherCrashOverride** [comment]
> JIT/CLR should know when a variable goes out of scope

Scope is a compiler/language concept.  The compiler must insert calls to an object's dtor when it sees a variable go out of scope.  This does not happen in C#.  Instead, the GC tracks objects "liveness" during collections in a non-deterministic way.  Only after a GC cycle will a reference count be known.

To make this work with C#, CoreCLR, Roslyn and the C# language will need to be modified. A reference count variable will need to be added to object headers. Dtors will need to be called when objects go out of scope and GC.SuppressFinalizer() after.  Additionally, assignment operators will need to be added to know when a copy of a variable is made.

TL;DR = This is not trivial. See also #430.

**Comment by mikedn** [comment]
> If rc is one, there's no need to copy, modifying the object directly is safe. 

This looks like a race condition. What stops another thread from incrementing the ref count after you determined that the ref count is 1?

> if(RC.get(this)>1)this=new test(); //should be allowed

What exactly is the meaning of `this = new test();`?

> c.add(1);//the reference of c changed, the reference of b not changed, c!=b, c.a=2, b.a=1

Apparently you're suggesting that `this` should be passed by ref instead of by value. Leaving aside the inherent issues associated with making such a change in an established runtime and language, what makes you think that a feature that changes variables in ways that aren't obvious to the developer is a good idea?

**Comment by ygc369** [comment]
@mikedn 
The Compiler and CLR shouldn't allow objects with rc accessed by multi-threads.
A new grammar should be created to obviously pass "this" by ref instead of by value.
for example:
c..add(1); //use ".." instead of "." to announce that the reference of c may be changed after this call.

**Comment by ayende** [comment]
That means a huge number of changes.
Or, you can have your own RC system, and rely on using() for them.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by svick** [comment]
@ygc369 I thought being thread-safe is one of the major advantages of immutable types.

**Comment by mikedn** [comment]
> The Compiler and CLR shouldn't allow objects with rc accessed by multi-threads.

This raises more questions than it answers. How can it be implemented? Is this feature still useful/usable/efficient after such a constraint is applied?

Part of the problem is that you're trying to use reference counting for something that it wasn't really designed for. In typical reference counting implementations the reference count is irrelevant except when it drops to 0. Nobody cares if the reference count is 1 or 42 and this allows reference counting to work reasonably well in a multithreaded application. The reference count is atomically incremented and thus it is possible for multiple threads to create copies of a reference safely and efficiently. With you requirements that's not longer possible and that's rather odd.

It may be possible to use reference counting to implement copy on write but it's very likely that the API needs to be quite different. Something like `var unique = RC.TryAcquireUniqueReference(ref variable)`. That will check if the ref count is one and if yes the reference stored in `variable` will be returned and `variable` will be set to null. Of course, all this needs to be done atomically. How something like this would interact with `ref this` I don't know. My guess is that not very well...

**Comment by GSPP** [comment]
Reference counts can become "stuck" in case of cycles. This is the reason we need GC actually.

Ref counting could work *most* of the time, though, and be used to avoid a lot of GC work. A GC is needed from time to time to delete dead cycles.

I think the CLR team should spend their time elsewhere.
